tracing/hooks.go

```diff
@@ -1,74 +1,103 @@
 package tracing
 
-// BaseplateHook allows you to inject functionality into the lifecycle of a
+// CreateServerSpanHook allows you to inject functionality into the lifecycle of a
 // Baseplate request.
-type BaseplateHook interface {
-	OnServerSpanCreate(span *Span) error
+type CreateServerSpanHook interface {
+	// OnCreateServerSpan is called after a server Span is first created by
+	// tracing.CreateServerSpan, before any OnPostStart Hooks are called.
+	//
+	// OnCreateServerSpan is the recommended place to register Hooks onto the
+	// server Span.
+	OnCreateServerSpan(span *Span) error
 }
 
-// SpanHook allows you to inject functionality into the lifecycle of Baseplate
-// Spans.
-type SpanHook interface {
-	OnCreateChild(child *Span) error
-	OnStart(span *Span) error
-	OnEnd(span *Span, err error) error
+// CreateChildSpanHook allows you to inject functionality into the creation of a
+// Baseplate span.
+type CreateChildSpanHook interface {
+	// OnCreateChild is called after a child Span is first created, before any
+	// OnPostStart Hooks are called.
+	//
+	// OnCreateChild is the recommended place to register Hooks onto the
+	// child Span.
+	OnCreateChild(parent, child *Span) error
+}
+
+// StartStopSpanHook allows you to inject functionality immediately after
+// starting a span and immediately before stopping a span.
+type StartStopSpanHook interface {
+	// OnPostStart is called after a child Span is created and the OnCreateChild
+	// Hooks on the parent Span are called.
+	OnPostStart(span *Span) error
+
+	// OnPreStop is called by Span.Stop, after setting any custom tags, but
+	// before the span is stopped, serialized, or published.
+	OnPreStop(span *Span, err error) error
+}
+
+// SetSpanTagHook allows you to inject functionality after setting a tag on a
+// span.
+type SetSpanTagHook interface {
+	// OnSetTag is called by Span.SetTag, after the tag is set on the Span.
 	OnSetTag(span *Span, key string, value interface{}) error
+}
+
+// AddSpanCounterHook allows you to inject functionality after adding a counter
+// on a span.
+type AddSpanCounterHook interface {
+	// OnAddCounter is called by Span.AddCounter, after the counter is updated
+	// on the Span.
 	OnAddCounter(span *Span, key string, delta float64) error
 }
 
 var (
-	hooks []BaseplateHook
+	hooks []CreateServerSpanHook
 )
 
-// RegisterBaseplateHook registers a Hook into the Baseplate request lifecycle.
+func isSpanHook(hook interface{}) bool {
```

Comment:
🔕 This function is probably worth exporting so users implementing a hook can use it in their tests.